src/economy.cpp

```diff
@@ -1527,7 +1527,7 @@ struct ReserveCargoAction {
 
 	bool operator()(Vehicle *v)
 	{
-		if (v->cargo_cap > v->cargo.RemainingCount()) {
+		if (v->cargo_cap > v->cargo.RemainingCount() && (st->owner != OWNER_NONE || st->town->exclusive_counter == 0 || st->town->exclusivity == v->owner)) {
```

Comment:
Could you turn these additional tests into its own function, used both here and below? The function can then be documented to explain what the test is doing.